designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -72,77 +72,134 @@ class DeploymentServiceImpl(
-  // TODO: refactor this into smaller steps, currently "validators" in this method have a lot of side effects and the whole thing is pretty confusing
   override def deployProcessAsync(
-      processIdWithName: ProcessIdWithName,
+      processId: ProcessIdWithName,
       savepointPath: Option[String],
       comment: Option[String]
   )(implicit user: LoggedUser, ec: ExecutionContext): Future[Future[Option[ExternalDeploymentId]]] = {
     val actionType = ProcessActionType.Deploy
-    validateDeploymentComment(comment).flatMap { validatedComment =>
-      checkCanPerformActionAndAddInProgressAction[CanonicalProcess](
-        processIdWithName,
+
+    for {
+      // 1. common for all commands/actions
+      validatedComment <- validateDeploymentComment(comment)
+      ctx <- prepareCommandContextWithAction[CanonicalProcess](
+        processId,
         actionType,
-        d => Some(d.processVersionId),
-        d => Some(d.processingType)
-      ).flatMap { case (processDetails, actionId, versionOnWhichActionIsDone, buildInfoProcessIngType) =>
-        validateBeforeDeploy(processDetails, actionId).transformWith {
-          case Failure(ex) =>
-            dbioRunner.runInTransaction(actionRepository.removeAction(actionId)).transform(_ => Failure(ex))
-          case Success(validationResult) =>
-            // we notify of deployment finish/fail only if initial validation succeeded
-            val deploymentFuture = runDeploymentActionWithNotifications(
-              actionType,
-              actionId,
-              processIdWithName,
-              versionOnWhichActionIsDone,
-              validatedComment,
-              buildInfoProcessIngType
-            ) {
-              dispatcher
-                .deploymentManagerUnsafe(processDetails.processingType)
-                .processCommand(
-                  RunDeploymentCommand(
-                    validationResult.processVersion,
-                    validationResult.deploymentData,
-                    validationResult.resolvedScenario,
-                    savepointPath
-                  )
-                )
-            }
-            Future.successful(deploymentFuture)
-        }
+        p => Some(p.processVersionId),
+        p => Some(p.processingType)
+      )
+      // 2. command specific section
+      deployedScenarioData <- prepareDeployedScenarioData(ctx.processDetails, ctx.actionId)
+      command = RunDeploymentCommand(
+        deployedScenarioData.processVersion,
+        deployedScenarioData.deploymentData,
+        deployedScenarioData.resolvedScenario,
+        savepointPath
+      )
+      // TODO: move validateBeforeDeploy before creating an action
+      result <- validateBeforeDeploy(ctx.processDetails, deployedScenarioData).transformWith {
+        case Failure(ex) =>
+          dbioRunner.runInTransaction(actionRepository.removeAction(ctx.actionId)).transform(_ => Failure(ex))
```

Comment:
Perhaps we could extract some name for this for readability?

Reply:
Fixed.